query/src/servers/http/v1/http_query_handlers.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::Ordering;
+use std::sync::Arc;
+use std::time::Duration;
+use std::time::Instant;
+
+use common_base::ProgressValues;
+use common_datavalues::DataSchemaRef;
+use common_exception::Result;
+use hyper::http::header;
+use poem::get;
+use poem::http::StatusCode;
+use poem::post;
+use poem::web::Data;
+use poem::web::Json;
+use poem::web::Path;
+use poem::web::Query;
+use poem::IntoResponse;
+use poem::Response;
+use poem::Route;
+use serde::Deserialize;
+use serde::Serialize;
+
+use crate::servers::http::v1::block_to_json::JsonBlockRef;
+use crate::servers::http::v1::query::execute_state::HttpQueryRequest;
+use crate::servers::http::v1::query::http_query::HttpQueryResponseInternal;
+use crate::servers::http::v1::query::result_data_manager::Wait;
+use crate::sessions::SessionManagerRef;
+
+pub fn make_page_uri(query_id: &str, page_no: usize) -> String {
+    format!("/v1/query/{}/page/{}", query_id, page_no)
+}
+
+pub fn make_state_uri(query_id: &str) -> String {
+    format!("/v1/query/{}", query_id)
+}
+
+pub fn make_delete_uri(query_id: &str) -> String {
+    format!("/v1/query/{}/kill?delete=true", query_id)
+}
+
+pub fn make_kill_uri(query_id: &str) -> String {
+    format!("/v1/query/{}/kill?delete=false", query_id)
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct QueryResponse {
+    pub id: Option<String>,
+    pub columns: Option<DataSchemaRef>,
+    pub data: JsonBlockRef,
+    pub next_uri: Option<String>,
+    pub state_uri: Option<String>,
+    pub kill_uri: Option<String>,
+    pub delete_uri: Option<String>,
+    // TODO(youngsofun): consider better response for error
+    // 1. another json format for request error
+    // 2. return 400 for some situation
+    // 3. more detail from ErrorCode (does not support Serialization))
+    // 4. use error code to ease the handling in client program
+    pub request_error: Option<String>,
```

Comment:
how about define a QueryError which implements an IntoResponse? like this:

```
query_cancel_handler() -> Result<impl IntoResponse, QueryError>
```

maybe it could get the error handling easier